lrauv_ignition_plugins/src/ScienceSensorsSystem.cc

```diff
@@ -606,6 +710,249 @@ void ScienceSensorsSystemPrivate::PclVectorToEigen(
+    interpolatorInds1, interpolatorSqrDists1, _interpolators1);
+  if (interpolatorInds1.size() < 4 || interpolatorSqrDists1.size() < 4)
+  {
+    ignwarn << "Could not find enough neighbors in 1st slice z = " << nnZ
+      << " for trilinear interpolation." << std::endl;
+    return;
+  }
+
+  // Step 2: exclude z slice of 1st NN from further searches.
+
+  // Remove all points in the z slice of the 1st NN, so that the 2nd NN will be
+  // found in another z slice.
+  // Set invert flag to get all but the depth slice.
+  pcl::PointCloud<pcl::PointXYZ> cloudExceptZSlice1;
+  std::vector<int> indsExceptZSlice1;
+  this->CreateDepthSlice(nnZ, *(this->timeSpaceCoords[this->timeIdx]),
+    cloudExceptZSlice1, indsExceptZSlice1, true);
+  igndbg << "Excluding 1st nn z slice. Remaining cloud has "
+    << cloudExceptZSlice1.points.size() << " points" << std::endl;
+
+  // Step 3: Look for 2nd NN everywhere except z slice of 1st NN.
+  // In this 2nd z-slice, search for 4 NNs
+
+  // Search for 2nd NN
+  std::vector<int> inds2;
+  std::vector<float> sqrDists2;
+  std::vector<pcl::PointXYZ> nbrs2;
+  this->CreateAndSearchOctree(_pt, cloudExceptZSlice1,
+    inds2, sqrDists2, nbrs2, 1);
+  if (inds2.size() < 1 || sqrDists2.size() < 1)
+  {
+    ignwarn << "Could not find 2nd NN among "
+      << cloudExceptZSlice1.points.size()
+      << " points for trilinear interpolation" << std::endl;
+    return;
+  }
+
+  // Take closest point as 2nd NN
+  int nnIdx2 = inds2[0];
+  float minDist2 = sqrDists2[0];
+  // Get z of neighbor
+  float nnZ2 = nbrs2[0].z;
+
+  // Step 4: Look for 4 NNs in the z slice of 2nd NN
+
+  // Search in z slice of 1st NN for 4 nearest neighbors in this slice
+  this->CreateDepthSlice(nnZ2, cloudExceptZSlice1, zSlice2, zSliceInds2);
+  igndbg << "2nd nn idx " << nnIdx2 << ", dist " << sqrt(minDist2)
+    << ", z slice " << zSlice2.points.size() << " points" << std::endl;
+  this->CreateAndSearchOctree(_pt, zSlice2,
+    interpolatorInds2, interpolatorSqrDists2, _interpolators2);
+  if (interpolatorInds2.size() < 4 || interpolatorSqrDists2.size() < 4)
+  {
+    ignwarn << "Could not find enough neighbors in 2nd slice z = " << nnZ2
+      << " for trilinear interpolation." << std::endl;
+    return;
+  }
+}
+
+/////////////////////////////////////////////////
+void ScienceSensorsSystemPrivate::CreateDepthSlice(
+  float _depth,
+  pcl::PointCloud<pcl::PointXYZ> &_cloud,
+  pcl::PointCloud<pcl::PointXYZ> &_zSlice,
+  std::vector<int> &_zSliceInds,
+  bool _invert)
+{
+  // Separate a z slice, i.e. points with z equal to that of 1st NN
+  // Pass in extract_removed_indices=true to get indices of removed points
+  pcl::PassThrough<pcl::PointXYZ> passThruFilter =
+    pcl::PassThrough<pcl::PointXYZ>(true);
+  passThruFilter.setInputCloud(_cloud.makeShared());
+  passThruFilter.setNegative(_invert);
+  passThruFilter.setFilterFieldName("z");
+  passThruFilter.setFilterLimits(_depth - 1e-6, _depth + 1e-6);
```

Comment:
`1e-6` is actually a very standard tolerance used in a lot of places in Ignition, to do an `a == b` equivalence check while catching floating point errors. Lines 859-860 essentially check for `z == _depth`.

Reply:
There's a potential debate about what the range should be. But it definitely should not be a magic number in the code here. It should be passed in at a higher level such that it has potential to be adjusted for the different dataset if required.

Reply:
I see that you put an alternative in #142. Can we break just the part about this comment off into its own PR? See https://github.com/osrf/lrauv/pull/142#issuecomment-1021920916 I would really like to base proposals for big changes on #127, where the interpolation algorithms are implemented, because even if the search in #83 gives us some results, we now know that it doesn't mean those neighbors will work with interpolation, because of the constraints and edge cases in the interpolation algorithms.